kvbc/test/v4blockchain/v4_blockchain_test.cpp

```diff
@@ -2503,6 +2506,147 @@ TEST_F(v4_kvbc, digest_checks) {
   ASSERT_NE(empty_digest, blockchain->calculateBlockDigest(max_block));
 }
 
+TEST_F(v4_kvbc, get_from_snapshot) {
+  // Add blocks:
+  {  // block 1
+    categorization::Updates updates;
+    categorization::BlockMerkleUpdates merkle_updates;
+    merkle_updates.addUpdate("merkle_key1", "merkle_value_block1");
+    merkle_updates.addUpdate("merkle_key2", "merkle_value_block1");
+    updates.add("merkle", std::move(merkle_updates));
+    ASSERT_EQ(blockchain->add(std::move(updates)), (BlockId)1);
+  }
+  {  // block 2
+    categorization::Updates updates;
+    categorization::BlockMerkleUpdates merkle_updates;
+    merkle_updates.addUpdate("merkle_key1", "merkle_value_block2");
+    merkle_updates.addUpdate("merkle_key3", "merkle_value_block2");
+    updates.add("merkle", std::move(merkle_updates));
+    ASSERT_EQ(blockchain->add(std::move(updates)), (BlockId)2);
+  }
+  {  // block 3
+    categorization::Updates updates;
+    categorization::BlockMerkleUpdates merkle_updates;
+    merkle_updates.addUpdate("merkle_key2", "merkle_value_block3");
+    merkle_updates.addUpdate("merkle_key3", "merkle_value_block3");
+    updates.add("merkle", std::move(merkle_updates));
+    ASSERT_EQ(blockchain->add(std::move(updates)), (BlockId)3);
+  }
+  {  // block 4
+    categorization::Updates updates;
+    categorization::BlockMerkleUpdates merkle_updates;
+    merkle_updates.addUpdate("merkle_key2", "merkle_value_block4");
+    updates.add("merkle", std::move(merkle_updates));
+    ASSERT_EQ(blockchain->add(std::move(updates)), (BlockId)4);
+  }
+  auto opt_val = blockchain->getFromSnapshot("merkle", "merkle_key2", 2);
+  ASSERT_TRUE(opt_val);
+  ASSERT_EQ(std::get<categorization::MerkleValue>(*opt_val).block_id, 1);
+  opt_val = blockchain->getFromSnapshot("merkle", "merkle_key3", 2);
+  ASSERT_TRUE(opt_val);
+  ASSERT_EQ(std::get<categorization::MerkleValue>(*opt_val).block_id, 2);
+  opt_val = blockchain->getFromSnapshot("merkle", "merkle_key1", 3);
+  ASSERT_TRUE(opt_val);
+  ASSERT_EQ(std::get<categorization::MerkleValue>(*opt_val).block_id, 2);
+  opt_val = blockchain->getFromSnapshot("merkle", "merkle_key3", 3);
+  ASSERT_TRUE(opt_val);
+  ASSERT_EQ(std::get<categorization::MerkleValue>(*opt_val).block_id, 3);
+  opt_val = blockchain->getFromSnapshot("merkle", "merkle_key1", 4);
+  ASSERT_TRUE(opt_val);
+  ASSERT_EQ(std::get<categorization::MerkleValue>(*opt_val).block_id, 2);
```

Comment:
As discussed, let's add a test case where the key history is enabled, but the version is read from the latest CF

Reply:
added